waltz-tools/src/main/java/com/wepay/waltz/tools/zk/ZooKeeperCli.java

```diff
@@ -186,13 +188,19 @@ private void listClusterInfoAndServerPartitionAssignments(ZooKeeperClient zkClie
             ListCluster.list(root, zkClient);
         }
 
-        private void listReplicaState(ZNode znode, Map<ReplicaId, ReplicaState> replicaState) {
+        private void listReplicaState(ZNode znode, Map<ReplicaId, ReplicaState> replicaState, Long storeSessionId) {
             System.out.println("store [" + znode + "] replica states:");
 
             if (replicaState.size() > 0) {
                 Map<ReplicaId, ReplicaState> sortedReplicaState = new TreeMap<>(replicaState);
                 for (Map.Entry<ReplicaId, ReplicaState> entry : sortedReplicaState.entrySet()) {
-                    System.out.println("  " + entry.getKey() + ", SessionId: " + entry.getValue().sessionId + ", closingHighWaterMark: " + ((entry.getValue().closingHighWaterMark) == ReplicaState.UNRESOLVED ? "UNRESOLVED" : entry.getValue().closingHighWaterMark));
+                    long replicaSessionId = entry.getValue().sessionId;
+                    long closingHighWaterMark = entry.getValue().closingHighWaterMark;
+                    boolean recoveryFinished = storeSessionId != null && replicaSessionId == storeSessionId && closingHighWaterMark == ReplicaState.UNRESOLVED;
```

Comment:
In the recovery process, the sessionId is first updated in partitionMetadata at the beginning, and each replica session's sessionId is updated to the latest sessionId after the recovery finished, and its closing HWM set to UNRESOLVED. So here I check for these two conditions for complete.